File: src/js_module/gif_helper.js
// GIF 画像出力ヘルパ（JavaScript版）。

const fs = require("node:fs");
const path = require("node:path");

function grayscale_palette() {
  const p = new Array(256 * 3);
  for (let i = 0; i < 256; i += 1) {
    p[i * 3] = i;
    p[i * 3 + 1] = i;
    p[i * 3 + 2] = i;
  }
  return p;
}

function lzwEncode(data, minCodeSize) {
  if (data.length === 0) {
    return [];
  }
  const clearCode = 1 << minCodeSize;
  const endCode = clearCode + 1;
  const codeSize = minCodeSize + 1;
  const out = [];

  let bitBuffer = 0;
  let bitCount = 0;

  function emit(code) {
    bitBuffer |= code << bitCount;
    bitCount += codeSize;
    while (bitCount >= 8) {
      out.push(bitBuffer & 0xff);
      bitBuffer >>>= 8;
      bitCount -= 8;
    }
  }

  emit(clearCode);
  for (let i = 0; i < data.length; i += 1) {
    emit(data[i]);
    emit(clearCode);
  }
  emit(endCode);
  if (bitCount > 0) {
    out.push(bitBuffer & 0xff);
  }
  return out;
}

function pushU16LE(out, v) {
  out.push(v & 0xff);
  out.push((v >>> 8) & 0xff);
}

function save_gif(outPath, width, height, frames, palette, delay_cs = 4, loop = 0) {
  if (palette.length !== 256 * 3) {
    throw new Error("palette must be 256*3 bytes");
  }
  const frameBytes = width * height;
  for (const frame of frames) {
    if (frame.length !== frameBytes) {
      throw new Error("frame size mismatch");
    }
  }

  const chunks = [];
  const writeByte = (v) => chunks.push(Buffer.from([v & 0xff]));
  const writeU16 = (v) => {
    const b = Buffer.allocUnsafe(2);
    b[0] = v & 0xff;
    b[1] = (v >>> 8) & 0xff;
    chunks.push(b);
  };
  const writeBytes = (arr) => chunks.push(Buffer.from(arr));

  chunks.push(Buffer.from("GIF89a", "ascii"));
  writeU16(width);
  writeU16(height);
  writeByte(0xf7);
  writeByte(0);
  writeByte(0);
  writeBytes(palette);

  writeBytes([0x21, 0xff, 0x0b]);
  chunks.push(Buffer.from("NETSCAPE2.0", "ascii"));
  writeBytes([0x03, 0x01]);
  writeU16(loop);
  writeByte(0x00);

  for (const frame of frames) {
    writeBytes([0x21, 0xf9, 0x04, 0x00]);
    writeU16(delay_cs);
    writeByte(0x00);
    writeByte(0x00);

    writeByte(0x2c);
    writeU16(0);
    writeU16(0);
    writeU16(width);
    writeU16(height);
    writeByte(0x00);

    writeByte(8);
    const compressed = lzwEncode(frame, 8);
    let pos = 0;
    while (pos < compressed.length) {
      const len = Math.min(255, compressed.length - pos);
      writeByte(len);
      writeBytes(compressed.slice(pos, pos + len));
      pos += len;
    }
    writeByte(0x00);
  }

  writeByte(0x3b);
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, Buffer.concat(chunks));
}

module.exports = { grayscale_palette, save_gif };

File: src/cs_module/gif_helper.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Pytra.CsModule
{
    // Python の pylib.gif_helper 相当実装。
    public static class gif_helper
    {
        private static void EmitCode(List<byte> outv, ref int bitBuffer, ref int bitCount, int code, int codeSize)
        {
            bitBuffer |= (code << bitCount);
            bitCount += codeSize;
            while (bitCount >= 8)
            {
                outv.Add((byte)(bitBuffer & 0xFF));
                bitBuffer >>= 8;
                bitCount -= 8;
            }
        }

        private static byte[] LzwEncode(IReadOnlyList<byte> data, int minCodeSize)
        {
            if (data.Count == 0)
            {
                return Array.Empty<byte>();
            }

            int clearCode = 1 << minCodeSize;
            int endCode = clearCode + 1;
            int codeSize = minCodeSize + 1;

            var outv = new List<byte>();
            int bitBuffer = 0;
            int bitCount = 0;

            EmitCode(outv, ref bitBuffer, ref bitCount, clearCode, codeSize);

            for (int i = 0; i < data.Count; i++)
            {
                EmitCode(outv, ref bitBuffer, ref bitCount, data[i], codeSize);
                EmitCode(outv, ref bitBuffer, ref bitCount, clearCode, codeSize);
            }

            EmitCode(outv, ref bitBuffer, ref bitCount, endCode, codeSize);

            if (bitCount > 0)
            {
                outv.Add((byte)(bitBuffer & 0xFF));
            }

            return outv.ToArray();
        }

        public static List<byte> grayscale_palette()
        {
            var p = new List<byte>(256 * 3);
            for (int i = 0; i < 256; i++)
            {
                byte v = (byte)i;
                p.Add(v);
                p.Add(v);
                p.Add(v);
            }
            return p;
        }

        private static void AppendU16LE(List<byte> outv, int v)
        {
            outv.Add((byte)(v & 0xFF));
            outv.Add((byte)((v >> 8) & 0xFF));
        }

        public static void save_gif(
            string path,
            long width,
            long height,
            List<List<byte>> frames,
            List<byte> palette,
            long delay_cs,
            long loop)
        {
            int w = checked((int)width);
            int h = checked((int)height);
            int delay = checked((int)delay_cs);
            int loopCount = checked((int)loop);

            if (palette.Count != 256 * 3)
            {
                throw new ArgumentException("palette must be 256*3 bytes");
            }
            int frameSize = checked(w * h);
            foreach (List<byte> fr in frames)
            {
                if (fr.Count != frameSize)
                {
                    throw new ArgumentException("frame size mismatch");
                }
            }

            var outv = new List<byte>(1024 + frames.Count * frameSize / 2);

            outv.Add((byte)'G'); outv.Add((byte)'I'); outv.Add((byte)'F'); outv.Add((byte)'8'); outv.Add((byte)'9'); outv.Add((byte)'a');
            AppendU16LE(outv, w);
            AppendU16LE(outv, h);
            outv.Add(0xF7); outv.Add(0x00); outv.Add(0x00);
            outv.AddRange(palette);

            outv.Add(0x21); outv.Add(0xFF); outv.Add(0x0B);
            outv.Add((byte)'N'); outv.Add((byte)'E'); outv.Add((byte)'T'); outv.Add((byte)'S'); outv.Add((byte)'C');
            outv.Add((byte)'A'); outv.Add((byte)'P'); outv.Add((byte)'E'); outv.Add((byte)'2'); outv.Add((byte)'.'); outv.Add((byte)'0');
            outv.Add(0x03); outv.Add(0x01);
            AppendU16LE(outv, loopCount);
            outv.Add(0x00);

            foreach (List<byte> fr in frames)
            {
                outv.Add(0x21); outv.Add(0xF9); outv.Add(0x04); outv.Add(0x00);
                AppendU16LE(outv, delay);
                outv.Add(0x00); outv.Add(0x00);

                outv.Add(0x2C);
                AppendU16LE(outv, 0);
                AppendU16LE(outv, 0);
                AppendU16LE(outv, w);
                AppendU16LE(outv, h);
                outv.Add(0x00);

                outv.Add(0x08);
                byte[] compressed = LzwEncode(fr, 8);
                int pos = 0;
                while (pos < compressed.Length)
                {
                    int len = Math.Min(255, compressed.Length - pos);
                    outv.Add((byte)len);
                    for (int i = 0; i < len; i++)
                    {
                        outv.Add(compressed[pos + i]);
                    }
                    pos += len;
                }
                outv.Add(0x00);
            }

            outv.Add(0x3B);
            File.WriteAllBytes(path, outv.ToArray());
        }

        public static void save_gif(
            string path,
            long width,
            long height,
            List<List<byte>> frames,
            List<byte> palette)
        {
            save_gif(path, width, height, frames, palette, 4, 0);
        }
    }
}
